### PR TITLE
Port `deduplicate` to TQL2

### PR DESCRIPTION
**Pull request by dominiklohmann**
This ports the `deduplicate` operator to TQL2. The operator has two major changes from its TQL1 pendant, namely:
- It takes only a single expression as an argument. To group by multiple keys, a record expression must be used.
- The `timeout` option was split into separate `create_timeout`, `write_timeout`, and `read_timeout` options. Previously, the timeout behaved like a read timeout, i.e., it always reset, even when the event was not forwarded.

**Example 1**: Get diagnostics, but show at most 10 per pipeline per hour:

```
diagnostics live=true
deduplicate pipeline_id, limit=10, create_timeout=1h
```

**Example 2**: Get an event every time the node disconnects from the Tenzir Platform:

```
metrics "platform", live=true
deduplicate connected, distance=1
where not connected
```

### CODE REVIEW

File: libtenzir/builtins/operators/deduplicate.cpp [hunk 1]
```diff
@@ -597,6 +601,186 @@ class plugin final : public virtual operator_plugin<deduplicate_operator> {
...
+  auto make(invocation inv, session ctx) const
+    -> failure_or<operator_ptr> override {
+    auto cfg = configuration2{};
+    auto parser = argument_parser2::operator_("deduplicate");
+    parser.positional("expr", cfg.expr_, "<expr>");
+    parser.named("distance", cfg.distance_);
+    parser.named("limit", cfg.limit_);
+    parser.named("create_timeout", cfg.create_timeout_);
+    parser.named("write_timeout", cfg.write_timeout_);
+    parser.named("read_timeout", cfg.read_timeout_);
```
Comment (dominiklohmann) [inline libtenzir/builtins/operators/deduplicate.cpp:759]: I still need to write docs, but before that I wanted to hear what @mavam thinks about separating `timeout` into three options. The TQL1 operator only had the equivalent of `read_timeout`.

The timeouts behave the following way:
- `create_timeout` is determined at the time when the first event for a key is let through
- `write_timeout` is determined whenever an event for a key is let through; if `limit` is 1, this is the same as `create_timeout`
- `read_timeout` is determined whenever a key is found, no matter whether an event was let through or not.

This allows saying things like `deduplicate foo, limit=10, create_timeout=1h` to let 10 events per unique `foo` through every hour, which was not possible with the TQL1 operator.

File: libtenzir/builtins/operators/deduplicate.cpp [hunk 2]
```diff
@@ -597,6 +601,186 @@ class plugin final : public virtual operator_plugin<deduplicate_operator> {
...
+public:
+  deduplicate_operator2() = default;
+
+  explicit deduplicate_operator2(configuration2 cfg) : cfg_(std::move(cfg)) {
+  }
+
+  auto
+  operator()(generator<table_slice> input, operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    tsl::robin_map<data, state2> state = {};
```
Comment (dominiklohmann) [inline libtenzir/builtins/operators/deduplicate.cpp:651]: The old deduplicate operator had an internal loop that cleaned up stale data with an expired distance or timeout. I still need to re-introduce that.

**Review by IyeOnline — APPROVED** [review]
LGTM

**Comment by raxyte** [comment]
> It takes only a single expression as an argument. To group by multiple keys, a record expression must be used.

If it's because the argument parser lacks support for variadic args, not sure this is a good idea.

**Comment by dominiklohmann** [comment]
> If it's because the argument parser lacks support for variadic args, not sure this is a good idea.

It is because of precisely that reason currently, we can always change this again in the future.

File: web/docs/tql2/operators/deduplicate.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
+# sample
```
Comment (mavam) [inline web/docs/tql2/operators/deduplicate.md:1]:
```suggestion
# deduplicate
```

File: web/docs/tql2/operators/deduplicate.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+## Description
+
+The `deduplicate` operator removes duplicates from a stream of events, based
+on the value of one or more fields.
+
+The operator has four tuning knobs:
+
+1. **Key**: The expression to deduplicate. Defaults to `this`, i.e.,
+   deduplicating entire events. To deduplicate multiple fields, use a record
+   expression.
```
[the remark this answers is not in the document]
Comment (dominiklohmann) [reply web/docs/tql2/operators/deduplicate.md]: Good idea, I've added one.

File: libtenzir/builtins/operators/deduplicate.cpp [hunk 5]
```diff
@@ -597,6 +601,231 @@ class plugin final : public virtual operator_plugin<deduplicate_operator> {
...
+class tql2_plugin final
+  : public virtual operator_plugin2<deduplicate_operator2> {
+public:
+  auto make(invocation inv, session ctx) const
+    -> failure_or<operator_ptr> override {
+    auto key = std::optional<ast::expression>{};
+    auto limit = std::optional<located<int64_t>>{};
+    auto cfg = configuration2{};
+    auto parser = argument_parser2::operator_("deduplicate");
+    parser.positional("key", key, "<expr>");
```
Comment (jachris) [inline libtenzir/builtins/operators/deduplicate.cpp:755]:
```suggestion
    parser.positional("key", key, "any");
```

File: web/docs/tql2/operators/deduplicate.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+2. **Limit**: The number of duplicate keys allowed before an event is
+   suppressed. Defaults to 1, which is equivalent to removing all duplicates.
+3. **Distance**: Distance between two events that can be considered duplicates.
+   A value of 1 means that only adjacent events can be considered duplicate.
+   When unspecified, the distance is infinite.
+4. **Timeout**: The time that needs to pass until a surpressed event is no
+   longer considered a duplicate. The create timeout resets when the first event
+   for a given key is let through, the write timeout resets when any event for a
+   given key is let through, and the read timeout is reset when a key is seen,
+   even if the event is suppressed.
```
Comment (jachris) [inline web/docs/tql2/operators/deduplicate.md:29]: This uses a very different format for documenting the arguments than we usually do. Not saying it's a bad thing (I was also inclined towards more free form docs), but do we want to keep it that way?

(I would not block the PR for this, but the other conversation will anyway)

Comment (dominiklohmann) [reply web/docs/tql2/operators/deduplicate.md:29]: I copied this from some other operator, and didn't think about it too deeply. Will align this.

Comment (dominiklohmann) [reply web/docs/tql2/operators/deduplicate.md]: Done and pushed.

**Pull request merged**